lightbeam/validate.py

```diff
@@ -108,33 +292,136 @@ def load_local_descriptors(self):
                 if not self.is_valid_descriptor_value(namespace, codeValue):
                     return payload[k] + f" is not a valid descriptor value for {k}" + (" (at " + path + ")" if path!="" else "")
         return ""
 
+    # Validates descriptor values for a single payload (returns an error message or empty string)
+    def has_invalid_references(self, payload, path=""):
+        for k in payload.keys():
+            if isinstance(payload[k], dict) and not k.endswith("Reference"):
+                value = self.has_invalid_references(payload[k], path+("." if path!="" else "")+k)
+                if value!="": return value
+            elif isinstance(payload[k], list):
+                for i in range(0, len(payload[k])):
+                    value = self.has_invalid_references(payload[k][i], path+("." if path!="" else "")+k+"["+str(i)+"]")
+                    if value!="": return value
+            elif isinstance(payload[k], dict) and k.endswith("Reference"):
+                is_valid_reference = False
+                original_endpoint = util.pluralize_endpoint(k.replace("Reference",""))
+
+                # this deals with the fact that an educationOrganizationReference may be to a school, LEA, etc.:
+                endpoints_to_check = self.EDFI_GENERICS_TO_RESOURCES_MAPPING.get(original_endpoint, [original_endpoint])
+                for endpoint in endpoints_to_check:
+                    # check if it's a local reference:
+                    if endpoint not in self.local_reference_cache.keys(): break
+                    # construct cache_key for reference
+                    cache_key = self.get_cache_key(payload[k])
+                    if cache_key in self.local_reference_cache[endpoint]:
+                        is_valid_reference = True
+                        break
+                if not is_valid_reference: # not found in local data...
+                    for endpoint in endpoints_to_check:
+                        # check if it's a remote reference:
+                        params = payload[k].copy()
+                        if "link" in params.keys(): del params["link"]
+                        value = self.remote_reference_exists(endpoint, params)
+                        if value:
+                            is_valid_reference = True
+                            break
+                    if not is_valid_reference:
+                        return f"payload contains an invalid {k} " + (" (at "+path+"): " if path!="" else ": ") + json.dumps(params)
+        return ""
+
     # Tells you if a specified descriptor value is valid or not
     def is_valid_descriptor_value(self, namespace, codeValue):
         for row in self.lightbeam.api.descriptor_values:
             if row[1]==namespace and row[2]==codeValue:
                 return True
         return False
+
+    @staticmethod
+    def get_cache_key(payload):
+        cache_key = ''
+        for k in payload.keys():
+            cache_key += f"{payload[k]}~~~"
```

Comment:
Might it be worth spinning this out into its own even smaller static method so that writing to and reading from the cache are always syntactically aligned? I'm not sure

Reply:
Are you pointing out that this is similar to the functionality in `references_data_to_cache()`? it is, and importantly, it's _slightly_ different - the keys are sorted alphabetically there but not here... which could lead to inconsistent cache keys for the same structure.

TBH there's probably a more performant way to build and reference this local cache, but I propose we leave that as a future refactor/improvement. Is that ok by you, @johncmerfeld ?

Reply:
Definitely not a blocker to merging!

Yeah I think the main thing I'm reacting to is just the way keys are constructed (`f"{payload[k]}~~~"`). Since that's a novel format it could be preferable for it to be further abstracted away from the developer. Totally fair that the two functions to use those keys in different ways; just wondering about a hypothetical future where the triple-tilde syntax is written in many more places.

Reply:
But very possibly a case of premature abstraction - like I say, it's totally good as is